### ISSUE TITLE
Behaviour of !! operator has changed in 11.1

### ISSUE THREAD

**Issue by stogle** [AvaloniaUI/Avalonia#16071]
### Describe the bug

The behaviour of the double-[negation operator](https://docs.avaloniaui.net/docs/reference/built-in-data-binding-converters#negation-operator-examples) (`!!`) has changed between 11.0.10 and 11.1.0-rc1.

Consider the binding `{Binding !!Value}`, where `Value` is `null`. In 11.0.10 this binding yields `false`. In 11.1.0-rc1 the same binding yields `null`. This is a breaking change for my application.

### To Reproduce

<details>
  <summary>MainWindow.axaml</summary>

```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:MultiBindingBug.ViewModels"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="MultiBindingBug.Views.MainWindow"
        x:DataType="vm:MainWindowViewModel">
    <StackPanel>
        <StackPanel Orientation="Horizontal">
            <TextBlock Text="Value1: " />
            <TextBlock Text="{Binding Value1}" />
        </StackPanel>
        <StackPanel Orientation="Horizontal">
            <TextBlock Text="!!Value1: " />
            <TextBlock Text="{Binding !!Value1}" />
        </StackPanel>

        <StackPanel Orientation="Horizontal">
            <TextBlock Text="Value2: " />
            <TextBlock Text="{Binding Value2}" />
        </StackPanel>
        <StackPanel Orientation="Horizontal">
            <TextBlock Text="!!Value2: " />
            <TextBlock Text="{Binding !!Value2}" />
        </StackPanel>
    </StackPanel>
</Window>
```
</details>

<details>
  <summary>MainWindowViewModel.cs</summary>

```cs
namespace MultiBindingBug.ViewModels;

public partial class MainWindowViewModel
{
    public object? Value1 { get; } = 0;
    public object? Value2 { get; } = null;
}
```
</details>

Steps to reproduce:
1. Run the application. Avalonia 11.1.0-rc1 displays:
![image](https://github.com/AvaloniaUI/Avalonia/assets/11137040/85f77c51-ff9e-4426-a997-80773ccddfee)

### Expected behavior

Avalonia 11.0.10 displays:
![image](https://github.com/AvaloniaUI/Avalonia/assets/11137040/20a803c0-63e5-4711-b00a-a5e64371fa14)

### Avalonia version

11.1.0-rc1

### OS

Windows

**Comment by maxkatz6** [comment]
"!!" wasn't a defined and documented feature before. I would recommend using more explicit [ObjectConverters.IsNotNull](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Data/Converters/ObjectConverters.cs#L17).

**Comment by stogle** [comment]
> "!!" wasn't a defined and documented feature before. I would recommend using more explicit [ObjectConverters.IsNotNull](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Data/Converters/ObjectConverters.cs#L17).

It was documented. From the [link](https://docs.avaloniaui.net/docs/reference/built-in-data-binding-converters#negation-operator-examples) I posted above:

"You can also use the negation operator twice. For example, where you want to perform the conversion from integer to Boolean, and then negate that value."

This is not equivalent to any of the `ObjectConverters`.

**Comment by maxkatz6** [comment]
> It was documented

TIL, thanks. Either way, I agree that it's a regression.

> This is not equivalent to any of the ObjectConverters.

True, I was referring specifically to the "null to boolean" conversion.

**Comment by grokys** [comment]
Definitely looks like a regression introduced by the binding system refactor, will investigate.

Not sure if this is a showstopper and needs to be fixed for 11.0.0 or whether it can wait for 11.0.1 though.

**Comment by grokys** [comment]
We've decided that this issue is a serious enough regression that it necessitates an rc2.

**Comment by stogle** [comment]
I've confirmed that this issue is fixed for me in 11.1.0-RC2.
